libraries/UHS_host/UHS_ADK/UHS_ADK.h

```diff
@@ -0,0 +1,133 @@
+
+#include <UHS_host.h>
+
+#define ADK_VID   0x18D1
+#define ADK_PID   0x2D00
+#define ADB_PID   0x2D01
+
+#define XOOM  //enables repeating getProto() and getConf() attempts
+//necessary for slow devices such as Motorola XOOM
+//defined by default, can be commented out to save memory
+
+/* requests */
+
+#define ADK_GETPROTO      51  // check USB accessory protocol version
+#define ADK_SENDSTR       52  // send identifying string
+#define ADK_ACCSTART      53  // start device in accessory mode
+
+#define bmREQ_ADK_GET     USB_SETUP_DEVICE_TO_HOST|USB_SETUP_TYPE_VENDOR|USB_SETUP_RECIPIENT_DEVICE
+#define bmREQ_ADK_SEND    USB_SETUP_HOST_TO_DEVICE|USB_SETUP_TYPE_VENDOR|USB_SETUP_RECIPIENT_DEVICE
+
+#define ACCESSORY_STRING_MANUFACTURER   0
+#define ACCESSORY_STRING_MODEL          1
+#define ACCESSORY_STRING_DESCRIPTION    2
+#define ACCESSORY_STRING_VERSION        3
+#define ACCESSORY_STRING_URI            4
+#define ACCESSORY_STRING_SERIAL         5
+
+#define ADK_MAX_ENDPOINTS 3 // endpoint 0, bulk_IN, bulk_OUT
+
+#if DEBUG_PRINTF_EXTRA_HUGE
+#ifdef DEBUG_PRINTF_EXTRA_HUGE_ADK_HOST
+#define ADK_HOST_DEBUG(...) printf(__VA_ARGS__)
+#else
+#define ADK_HOST_DEBUG(...) VOID0
+#endif
+#else
+#define ADK_HOST_DEBUG(...) VOID0
+#endif
+
+class ADK : public UHS_USBInterface {
+protected:
+        /* ID strings */
+        const char* manufacturer;
+        const char* model;
+        const char* description;
+        const char* version;
+        const char* uri;
+        const char* serial;
+        uint16_t adkproto;
+
+
+        volatile bool ready; // device ready indicator
+
+        /* ADK proprietary requests */
+        uint8_t getProto(uint8_t* adkproto);
+        uint8_t sendStr(uint8_t index, const char* str);
+        uint8_t switchAcc(void);
+
+public:
+        static const uint8_t epDataInIndex = 1; // DataIn endpoint index
+        static const uint8_t epDataOutIndex = 2; // DataOUT endpoint index
+        volatile UHS_EpInfo epInfo[ADK_MAX_ENDPOINTS];
+
+        ADK(UHS_USB_HOST_BASE *p, const char* manufacturer,
+                const char* model,
+                const char* description,
+                const char* version,
+                const char* uri,
+                const char* serial);
+
+        // Methods for receiving and sending data
+        uint8_t Read(uint16_t *nbytesptr, uint8_t *dataptr);
+        uint8_t Write(uint16_t nbytes, uint8_t *dataptr);
+
+        uint8_t OnStart(void);
+        bool OKtoEnumerate(ENUMERATION_INFO *ei);
+        uint8_t SetInterface(ENUMERATION_INFO *ei);
+        void DriverDefaults(void);
+
+        uint8_t GetAddress() {
+                return bAddress;
+        };
+
+        bool isReady() {
+                pUsb->DisablePoll();
+                bool rv = ready;
+                pUsb->EnablePoll();
+                return rv;
+        };
+};
+
+/* get ADK protocol version */
+
+/* returns 2 bytes in *adkproto */
+inline uint8_t ADK::getProto(uint8_t* adkproto) {
```

Comment:
Don't use the inline keyword. This ends up actually bloating the code.
Instead use UHS_NI
For an example see here:
[https://github.com/felis/UHS30/blob/master/libraries/UHS_host/UHS_CDC_ACM/UHS_CDC_ACM_INLINE.h#L23](url)